keps/sig-auth/20190607-certificates-api.md

```diff
@@ -132,6 +133,75 @@ A typical successful issuance proceeds as follows.
    the `.Status.Certificate` field.
 1. The requestor observes the update, and stores the certificate locally.
 
+### Signers
+CSRs have a `signerName` field which is used to specify which signer the CSR creator wants to sign the certificate.
+To support migration from v1beta1 to v1, this required field will be defaulted in v1beta1 (optional in openapi), but
+not defaulted and required in v1 :
+ 1. If it's a kubelet client certificate, it is assigned "kubernetes.io/kubelet-client".
+ 2. Otherwise, it is assigned "kubernetes.io/legacy-unknown".
+
+There will be field selector support to make approvers and signers easier to write.
+
+All signers should provide information about how they work so that clients can predict what will happen to their CSRs.
+This includes:
+ 1. Trust distribution - how trust (ca bundles) are distributed.
+ 2. Permitted subjects - (any? specific subtree?) and behavior when a disallowed subject is requested.
+ 3. Permitted x509 extensions - (IP SANs? DNS SANs? Email SANs? URI SANs? others?) and behavior when a disallowed
+ extension is requested.
+ 4. Permitted key usages / extended key usages - (client only? server only? any? signer-determined? CSR-determined?) and
+ behavior when usages different than the signer-determined usages are specified in the CSR.
+ 5. Expiration/cert lifetime - (fixed by signer? configurable by admin? CSR-determined?) and behavior when an expiration
+ different than the signer-determined expiration is specified in the CSR.
+ 6. CA bit allowed/disallowed - and behavior if a CSR contains a request a for a CA cert when the signer does not permit it.
+
+
+sig-auth reserves all `kubernetes.io/*` `signerNames` and more may be added in the future.
+Kubernetes provides the following well-known signers.  Today, failures for all of these are only reported in kube-controller-manager logs:
+ 1. kubernetes.io/kube-apiserver-client - signs certificates that will be honored as client-certs by the kube-apiserver.
+    Never auto-approved by kube-controller-manager.
+    1. Trust distribution: signed certificates must be honored as client-certificates by the kube-apiserver.  The CA bundle
+       is not distributed by any other means.
+    2. Permitted subjects - no subject restrictions.
```

Comment:
you would allow any/all of the following?
* O=system:masters
* O=system:unauthenticated
* CN=system:anonymous

Reply:
> you would allow any/all of the following?
> 
> * O=system:masters
> * O=system:unauthenticated
> * CN=system:anonymous

I would.  I would allow this to sign anything if the CSR were approved.

Reply:
Could signers choose not to issue those?

Reply:
Being able to sign for `O=system:masters` makes me nervous.  I can see arguments in both directions, but none of them make me feel comfortable.

Reply:
> Being able to sign for `O=system:masters` makes me nervous. I can see arguments in both directions, but none of them make me feel comfortable.

A distributor or cluster-admin can create any approver they want and any admission plugin they want to restrict it further.  If it's banned at a spec level, then no cluster can ever get the functionality.  The restriction use-case is always is always possible, but someone who wants to sign any client-cert cannot restrict their way to more power.

Reply:
talked with @deads2k about this and settled on an admission plugin to limit system:masters client cert requests specifically (which deployers can turn off if they want), and the following guidance:

> Certain subjects like cluster-admin level users or groups vary between distributions and installations, but deserve additional scrutiny before approval and signing.  An admission plugin is available to restrict system:masters, but it is often not the only cluster-admin subject in a cluster.